compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpCompilerPluginUtil.java

```diff
@@ -192,22 +192,58 @@ private static void validateArrayElementTypeInReturnType(SyntaxNodeAnalysisConte
             TypeDescKind typeDescKind = retrieveEffectiveTypeDesc(typeDescriptor);
-            if (typeDescKind != TypeDescKind.RECORD && typeDescKind != TypeDescKind.MAP &&
-                    typeDescKind != TypeDescKind.TABLE) {
+            if (typeDescKind == TypeDescKind.OBJECT || typeDescKind == TypeDescKind.ERROR) {
                 reportInvalidReturnType(ctx, node, typeStringValue, diagnosticCode);
+            } else if (typeDescKind == TypeDescKind.UNION) {
+                List<TypeSymbol> typeSymbols = ((UnionTypeSymbol) typeDescriptor).userSpecifiedMemberTypes();
+                for (TypeSymbol typeSymbol : typeSymbols) {
+                    TypeDescKind effectiveTypeDesc = retrieveEffectiveTypeDesc(typeSymbol);
+                    if (effectiveTypeDesc == TypeDescKind.OBJECT || effectiveTypeDesc == TypeDescKind.ERROR) {
+                        reportInvalidReturnType(ctx, node, typeStringValue, diagnosticCode);
+                        return;
+                    } else if (effectiveTypeDesc == TypeDescKind.TYPE_REFERENCE) {
+                        if (isStatusCodeResponseType((TypeReferenceTypeSymbol) typeSymbol)) {
+                            reportInvalidReturnType(ctx, node, typeStringValue, diagnosticCode);
+                            return;
+                        }
+                        Optional<String> definitionName = ((TypeReferenceTypeSymbol) typeSymbol).definition().getName();
+                        if (definitionName.isPresent() && definitionName.get().equals("StatusCodeResponse")) {
+                            reportInvalidReturnType(ctx, node, typeStringValue, diagnosticCode);
+                            return;
+                        }
+                        validateArrayElementTypeInReturnType(ctx, node, typeStringValue, typeSymbol, diagnosticCode);
+                    } else {
+                        validateReturnType(ctx, node, typeStringValue, typeSymbol, diagnosticCode, false);
+                    }
+                }
+            } else {
+                validateReturnType(ctx, node, typeStringValue, typeDescriptor, diagnosticCode, false);
             }
         } else if (kind == TypeDescKind.ARRAY) {
             memberTypeDescriptor = ((ArrayTypeSymbol) memberTypeDescriptor).memberTypeDescriptor();
-            validateArrayElementTypeInReturnType(ctx, node, typeStringValue, memberTypeDescriptor, diagnosticCode);
+            validateArrayElementTypeInReturnType(ctx, node, typeStringValue, memberTypeDescriptor,
+                    diagnosticCode);
         } else {
             reportInvalidReturnType(ctx, node, typeStringValue, diagnosticCode);
         }
     }
 
+    public static boolean isStatusCodeResponseType(TypeReferenceTypeSymbol typeSymbol) {
+        if (typeSymbol.getName().isPresent()) {
+            return typeSymbol.getName().get().equals("StatusCodeResponse");
```

Comment:
Also we should check whether this `StatusCodeResponse` record is from `http` package. So can't we reuse the `isHttpModuleType` function here?

Reply:
Updated with the suggestion